### PR TITLE
Fixed quantile nan propagation and implemented nanquantile

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44497 (nan)quantile CPU optimization
* **#44393 Fixed quantile nan propagation and implemented nanquantile**

torch.quantile now correctly propagates nan and implemented torch.nanquantile similar to numpy.nanquantile.

Differential Revision: [D23649613](https://our.internmc.facebook.com/intern/diff/D23649613)

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -4610,11 +4610,11 @@ def merge_dicts(*dicts):
 Returns the q-th quantiles of each row of the :attr:`input` tensor along the dimension
 :attr:`dim`, doing a linear interpolation when the q-th quantile lies between two
 data points. By default, :attr:`dim` is `None` resulting in the :attr:`input` tensor
-beign flattened before computation.
+being flattened before computation.
 
 If :attr:`q` is a 1D tensor, the first dimension of the result corresponds to the quantiles
```
Comment (mruberry) [inline torch/_torch_docs.py:4615]: "corresponds to the quantiles" and "what remains from the reduction" are vague expressions. Can we be more precise here?

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
     torch.Size([3, 2, 1])
 """.format(**single_dim_common))
 
+add_docstr(torch.nanquantile,
+           r"""
+nanquantile(input, q, dim=None, keepdim=False, *, out=None) -> Tensor
+
+This is a variant of :func:`torch.quantile` that ignores `NaN` values when
```
Comment (mruberry) [inline torch/_torch_docs.py:4656]: ... that "ignores" NaN values, computing the quantiles :attr:\`q\` as if NaN values in :attr:\`input\` did not exist. See the documentation for :func:\`torch.quantile\` and the example below.

I don't think you need to mention when NaN is produced in the output, but if you do I think it's more consistent to use terminology like "if all values in a reduced row are NaN..."

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
...
+    >>> a = torch.tensor([[float('nan'), float('nan')], [1, 2]])
+    >>> a
+    tensor([[nan, nan],
+            [1., 2.]])
+    >>> a.nanquantile(0.5, 0)
+    tensor([1., 2.])
+    >>> a.nanquantile(0.5, 1)
+    tensor([   nan, 1.5000])
+
+See :func:`torch.quantile`
```
Comment (mruberry) [inline torch/_torch_docs.py:4676]: Kill this line here. Typically we don't put anything after the example(s)

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
...
+then will the result be `NaN`.
+
+Example::
+
+    >>> torch.tensor([float('nan'), 1, 2]).quantile(0.5)
+    tensor(nan)
+    >>> torch.tensor([float('nan'), 1, 2]).nanquantile(0.5)
+    tensor(1.5000)
+
+    >>> a = torch.tensor([[float('nan'), float('nan')], [1, 2]])
```
Comment (mruberry) [inline torch/_torch_docs.py:4667]: nit: t =

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -4649,6 +4649,33 @@ def merge_dicts(*dicts):
...
+    >>> a
+    tensor([[nan, nan],
+            [1., 2.]])
+    >>> a.nanquantile(0.5, 0)
```
Comment (mruberry) [inline torch/_torch_docs.py:4671]: nit: add the dim= explicitly for clarity

**Review by mruberry — APPROVED** [review]
Looks good,
